daemon/c_seccomp.c

```diff
@@ -710,6 +810,126 @@ c_seccomp_handle_notify(int fd, unsigned events, UNUSED event_io_t *io, void *da
+		 *
+		 * We therefore use an offset +1 for the data.args array to
+		 * retrieve syscall parameters from seccomp req in case of
+		 * mknodat() and set dirfd to AT_FDCWD in case of mknod()
+		 */
+		if (req->data.nr == SYS_mknodat) {
+			dirfd = req->data.args[arg_offset];
+			arg_offset++;
+		} else { // SYS_mknod
+			dirfd = AT_FDCWD;
+		}
+
+		mode_t mode = req->data.args[1 + arg_offset];
+		dev_t dev = req->data.args[2 + arg_offset];
+
+		/* Check cap of target pid in its namespace */
+		if (!c_seccomp_capable(req->pid, CAP_MKNOD)) {
+			ERROR("Missing CAP_MKNOD for process %d!", req->pid);
+			goto out;
+		}
+
+		/* We only handle char and block devices, due to our seccomp filter */
+		char dev_type = S_ISCHR(mode) ? 'c' : 'b';
+
+		/* Check if dev is allowed (c_cgroups submodule) */
+		if (!container_is_device_allowed(seccomp->container, dev_type, major(dev),
+						 minor(dev))) {
+			ERROR("Missing cgroup permission for device (%c %d:%d) in process %d!",
+			      dev_type, major(dev), minor(dev), req->pid);
+			goto out;
+		}
+
+		int pathname_max_len = PATH_MAX;
+		char *pathname = mem_alloc0(pathname_max_len);
+		if (!(pathname = (char *)c_seccomp_fetch_vm_new(
+			      seccomp, req->pid, (void *)req->data.args[0 + arg_offset],
+			      pathname_max_len))) {
+			ERROR_ERRNO("Failed to fetch pathname string");
+			mem_free0(pathname);
+			goto out;
+		}
+
+		int cml_dirfd = AT_FDCWD;
+		if (dirfd != AT_FDCWD) {
+			int pidfd;
+			if (-1 == (pidfd = pidfd_open(req->pid, 0))) {
+				ERROR_ERRNO("Could not open pidfd for emulating %s()",
+					    syscall_name);
+				mem_free0(pathname);
+				goto out;
+			}
+
+			cml_dirfd = pidfd_getfd(pidfd, dirfd, 0);
+			if (cml_dirfd < 0) {
+				ERROR_ERRNO(
+					"Could not open dirfd in target process for emulating %s()",
+					syscall_name);
+				mem_free0(pathname);
+				goto out;
+			}
+		}
+
+		char *cwd = proc_get_cwd_new(req->pid);
+
+		DEBUG("Emulating %s by executing mknodat %s on behalf of container", syscall_name,
+		      pathname);
+
+		struct mknodat_fork_data mknodat_params = { .dirfd = cml_dirfd,
+							    .pathname = pathname,
+							    .cwd = cwd,
+							    .mode = mode,
+							    .dev = dev,
+							    .uid = container_get_uid(
+								    seccomp->container) };
+		if (-1 ==
+		    (ret_mknodat = namespace_exec(req->pid, CLONE_NEWNS, false,
+						  c_seccomp_do_mknodat_fork, &mknodat_params))) {
+			audit_log_event(NULL, FSA, CMLD, CONTAINER_ISOLATION,
+					"seccomp-emulation-failed",
+					compartment_get_name(seccomp->compartment), 2, "syscall",
+					SYS_mknodat);
+			ERROR_ERRNO("Failed to execute mknodat");
+			if (cwd)
+				mem_free0(cwd);
+			mem_free0(pathname);
+			if (cml_dirfd >= 0)
+				close(cml_dirfd);
+			goto out;
+		}
+
+		DEBUG("mknodat returned %d", ret_mknodat);
+
+		// prepare answer
+		resp->id = req->id;
+		resp->error = 0;
+		resp->val = ret_mknodat;
+
+		if (cwd)
+			mem_free0(cwd);
+		mem_free0(pathname);
```

Comment:
there is only one "goto out:" statement without freeing pathname
=> may move it before this statement, define a per-branch "out-mknod" label and free once at this label?